File: Easy/0021_merge-two-sorted-lists/merge-two-sorted-lists.go
package main

import (
	"fmt"
)

type ListNode struct {
	Val int
	Next *ListNode
}

func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	dummy := &ListNode{}
	tail := dummy

	for list1 != nil && list2 != nil {
		if (list1.Val <= list2.Val) {
			tail.Next = list1
			list1 = list1.Next
		} else {
			tail.Next = list2
			list2 = list2.Next
		}
		tail = tail.Next
	}

	tail.Next = list1
	if tail.Next == nil {
		tail.Next = list2
	}
	return dummy.Next
}

func buildLinkedList(values []int) *ListNode {
	dummy := &ListNode{}
	current := dummy
	for _, v := range values {
		current.Next = &ListNode{Val: v}
		current = current.Next
	}
	return dummy.Next
}

func printLinkedList(node *ListNode) {
	for node != nil {
		fmt.Printf("%d => ", node.Val)
		node = node.Next
	}
	fmt.Println("nil")
}

type Unit struct {
	l1 []int
	l2 []int
}

func main() {
	tests := []Unit{
		{l1: []int{1, 2, 4}, l2: []int{1, 3, 4}},
		{l1: []int{}, l2: []int{1}},
		{l1: []int{1}, l2: []int{}},
		{l1: []int{}, l2: []int{}},
	}

	for _, t := range tests {
		l1 := buildLinkedList(t.l1)
		l2 := buildLinkedList(t.l2)

		fmt.Println("---")
		fmt.Println("Input 1:", t.l1)
		fmt.Println("Input 2:", t.l2)
		fmt.Println("Result:")
		res := mergeTwoLists(l1, l2)
		printLinkedList(res)
	}
}

File: Easy/0021_merge-two-sorted-lists/merge-two-sorted-lists.py
#!/bin/python3

from typing import Optional, List

class ListNode:
    def __init__(self, val:int = 0, next: Optional['ListNode'] = None):
        self.val = val
        self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode()
        tail = dummy

        while list1 and list2:
            if list1.val <= list2.val:
                tail.next = list1
                list1 = list1.next
            else:
                tail.next = list2
                list2 = list2.next
            tail = tail.next

        tail.next = list1 if list1 else list2

        return dummy.next

def build_linked_list(values: List[int]) -> Optional[ListNode]:
    dummy = ListNode()
    current = dummy
    for v in values:
        current.next = ListNode(v)
        current = current.next
    return dummy.next

def print_linked_list(node: Optional[ListNode]) -> None:
    while (node):
        print(node.val, end=" -> ")
        node = node.next
    print ("None")

class Unit:
    def __init__(self, l1: List[int], l2: List[int]):
        self.l1 = l1
        self.l2 = l2

def main():
    tests: List[Unit] = {
        Unit([1,2,4], [1,3,4]),
        Unit([], [1]),
        Unit([1], []),
        Unit([], []),
    }
    for t in tests:
        l1 = build_linked_list(t.l1)
        l2 = build_linked_list(t.l2)
        res = Solution().mergeTwoLists(l1, l2)
        print("---")
        print("Input 1:", t.l1)
        print("Input 1:", t.l2)
        print("res:")
        print_linked_list(res)


if __name__ == "__main__":
    main()
